labcontrol/gui/static/js/plateViewer.js

```diff
@@ -479,6 +483,57 @@ PlateViewer.prototype.modifyWell = function(row, col, content) {
   });
 };
 
+/**
+ *
+ * Modify the contents of a well
+ *
+ * @param {int} row The row of the well being modified
+ * @param {int} col The column of the well being modified
+ * @param {string} content The new content of the well
+ *
+ **/
+PlateViewer.prototype.modifyWell = function(row, col, content) {
+  var that = this,
+    studyID = this.getActiveStudy();
+
+  // Perform automatic matching. Note that this is currently done every time
+  // modifyWell() is called, even if the user types in something like "blank"
+  // or even if the user selects something from a dropdown list. (That later
+  // case could probably be detected in order to avoid doing matching here.)
+  //
+  // I expect the actual matching to be pretty fast (assuming that there
+  // aren't thousands of active samples); the main bottleneck is how requests
+  // are made to the server from modifyWell() and patchWell() instead of in
+  // batch operations.
+  var possiblyNewContent = content;
+  // TODO: cache list of active samples so that we don't have to make this
+  // particular request every time modifyWell() is called.
+  get_active_samples().then(
+    function(sampleIDs) {
+      // If there is *exactly one* match with an active sample ID, use
+      // that instead. In any other case (0 matches or > 1 matches),
+      // manual resolution is required -- so we don't bother changing the
+      // cell content.
+      var matchingSamples = getSubstringMatches(content, sampleIDs);
+      if (matchingSamples.length === 1) {
+        possiblyNewContent = matchingSamples[0];
+        safeArrayDelete(that.wellClasses[row][col], "well-indeterminate");
+      } else if (matchingSamples.length > 1) {
+        addIfNotPresent(that.wellClasses[row][col], "well-indeterminate");
+      } else {
+        safeArrayDelete(that.wellClasses[row][col], "well-indeterminate");
+      }
+      that.patchWell(row, col, possiblyNewContent, studyID);
```

Comment:
`possiblyNewContent` might have been a bad choice for the variable name, sorry—even if this code fails to find a one-to-one match, it still updates the database with the `content` that was passed in (which also should be "new," since `modifyWell()` is called when a cell changes). I'm open to renaming this / adding comments to improve readability.

Reply:
Renamed `possiblyNewContent` to `possiblyMatchedContent`, in the hopes of making this a bit clearer. Can do more if needed.

Reply:
Renamed `possiblyNewContent` to `possiblyMatchedContent`, in the hopes of making this a bit clearer. Can do more if needed.